File: source/downloads/code/LeetCode/Python/Combination-Sum3.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        ret = []

        self.findSol(k,n,ret,[],1)
        return ret

    def findSol(self,k,n,ret,tmp,val):
        if n < 0:
            return
        if k == 0:
            if n != 0:
                return
            else:
                ret.append(tmp[:])
                return
        for i in range(val,10):
            tmp.append(i)
            self.findSol(k-1,n-i,ret,tmp,i+1)
            tmp.pop()
        return

File: downloads/code/LeetCode/Java/Combination-Sum3.java

public class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
            List<List<Integer>> ret= new ArrayList<List<Integer>> ();
            List<Integer> tmp = new ArrayList<Integer>();
            dfs(ret, tmp, k, n, 1);
            return ret;
        }

    public void dfs(List<List<Integer>> ret, List<Integer> tmp, int k, int n, int start){
        if(n==0 && tmp.size() == k){
            List<Integer> li = new ArrayList<Integer>(tmp);
            ret.add(li);
            return;
        }
        for(int i=start; i<=9; i++){
            tmp.add(i);
            dfs(ret, tmp, k, n-i,i+1);
            tmp.remove(tmp.size()-1);
        }
       
    }
}

